### DISCUSSION TITLE
Return Type of `.on_conflict(...).do_update().set(...)`

### DISCUSSION THREAD

**Discussion by coderPaddyS** [diesel-rs/diesel#4080] [Q&A]
Hey!
I want to compose an application using diesel and want to prepare statements without specifying the return clause or the connection type, such that different modules can use sync or async diesel to construct and execute the queries.

Now I stumbled on `.on_conflict(...).do_update().set(...)`. It seems to be not possible to construct the return type of this clause as the module upsert is private and does not reexport the required types.
And it seems to me there is no helper type to represent this return type either.

How could the return type of the following snippet look like?
```rust
pub struct ItemType {
    pub type_id: i32,
    pub type_name: String
}

impl ItemType {
        pub fn insert_or_update<DB: Backend>(&self) -> _ {
            diesel::insert_into(item_types::table)
                .values(self)
                .on_conflict(dsl::type_id)
                .do_update()
                .set(dsl::type_name.eq(&self.type_name))
        }
}
```

**Comment by weiznich** [comment]
We do currently not expose the necessary types for specifying that return type, so that's something that is not supported yet. 

We might accept a PR that changes this as long as it would allow changing the internal DSL structure later (so using type defs as for almost all other DSL methods) and as long as it works with the `#[auto_type]` attribute.
